lib/libmbedtls/core/tee_lmd_provider.c

```diff
@@ -949,41 +1021,313 @@ void crypto_cipher_final(void *ctx __unused, uint32_t algo __unused)
+	case TEE_ALG_DES3_CBC_MAC_PKCS5:
+		res = TEE_ERROR_NOT_SUPPORTED;
+		goto err;
+#endif
+#if defined(CFG_CRYPTO_CMAC)
+	case TEE_ALG_AES_CMAC:
+		res = TEE_ERROR_NOT_SUPPORTED;
+		goto err;
+#endif
+	default:
+		res = TEE_ERROR_NOT_SUPPORTED;
+		goto err;
+	}
+
+	*ctx_ret = ctx;
+	return TEE_SUCCESS;
+err:
+	free(ctx);
+	return res;
+}
+
+void crypto_mac_free_ctx(void *ctx, uint32_t algo __maybe_unused)
+{
+	size_t ctx_size __maybe_unused;
+
+	/*
+	 * Check that it's a supported algo, or crypto_mac_alloc_ctx()
+	 * could never have succeded above.
+	 */
 	if (ctx)
-		assert(0);
+		assert(!mac_get_ctx_size(algo, &ctx_size));
+
+	switch (algo) {
+#if defined(CFG_CRYPTO_HMAC)
+	case TEE_ALG_HMAC_MD5:
+	case TEE_ALG_HMAC_SHA224:
+	case TEE_ALG_HMAC_SHA1:
+	case TEE_ALG_HMAC_SHA256:
+	case TEE_ALG_HMAC_SHA384:
+	case TEE_ALG_HMAC_SHA512:
+		mbedtls_md_free(ctx);
+		break;
+#endif
+#if defined(CFG_CRYPTO_CBC_MAC)
+	case TEE_ALG_AES_CBC_MAC_NOPAD:
+	case TEE_ALG_AES_CBC_MAC_PKCS5:
+	case TEE_ALG_DES_CBC_MAC_NOPAD:
+	case TEE_ALG_DES_CBC_MAC_PKCS5:
+	case TEE_ALG_DES3_CBC_MAC_NOPAD:
+	case TEE_ALG_DES3_CBC_MAC_PKCS5:
+		break;
+#endif
+#if defined(CFG_CRYPTO_CMAC)
+	case TEE_ALG_AES_CMAC:
+		break;
+#endif
+	default:
+		break;
+	}
+	free(ctx);
 }
 
-void crypto_mac_copy_state(void *dst_ctx __unused, void *src_ctx __unused,
-			   uint32_t algo __unused)
+void crypto_mac_copy_state(void *dst_ctx, void *src_ctx, uint32_t algo)
 {
+#if defined(CFG_CRYPTO_HMAC) || defined(CFG_CRYPTO_CMAC)
+	int lmd_res;
+#endif
+
+	switch (algo) {
+#if defined(CFG_CRYPTO_HMAC)
+	case TEE_ALG_HMAC_MD5:
+	case TEE_ALG_HMAC_SHA224:
+	case TEE_ALG_HMAC_SHA1:
+	case TEE_ALG_HMAC_SHA256:
+	case TEE_ALG_HMAC_SHA384:
+	case TEE_ALG_HMAC_SHA512:
+		lmd_res = mbedtls_md_clone(dst_ctx, src_ctx);
+		if (lmd_res != 0)
+			FMSG("hmac clone failed, res is 0x%x", -lmd_res);
+		break;
+#endif
+#if defined(CFG_CRYPTO_CBC_MAC)
+	case TEE_ALG_AES_CBC_MAC_NOPAD:
+	case TEE_ALG_AES_CBC_MAC_PKCS5:
+	case TEE_ALG_DES_CBC_MAC_NOPAD:
+	case TEE_ALG_DES_CBC_MAC_PKCS5:
+	case TEE_ALG_DES3_CBC_MAC_NOPAD:
+	case TEE_ALG_DES3_CBC_MAC_PKCS5:
+		break;
+#endif
+#if defined(CFG_CRYPTO_CMAC)
+	case TEE_ALG_AES_CMAC:
+		break;
+#endif
+	default:
+		break;
+	}
 }
 
-TEE_Result crypto_mac_init(void *ctx __unused, uint32_t algo __unused,
-			   const uint8_t *key __unused, size_t len __unused)
+TEE_Result crypto_mac_init(void *ctx, uint32_t algo, const uint8_t *key,
+			   size_t len)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	int lmd_res __maybe_unused;
+
+	if (!ctx)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	switch (algo) {
+#if defined(CFG_CRYPTO_HMAC)
+	case TEE_ALG_HMAC_MD5:
+	case TEE_ALG_HMAC_SHA224:
+	case TEE_ALG_HMAC_SHA1:
+	case TEE_ALG_HMAC_SHA256:
+	case TEE_ALG_HMAC_SHA384:
+	case TEE_ALG_HMAC_SHA512:
+		mbedtls_md_hmac_reset(ctx);
+		lmd_res = mbedtls_md_hmac_starts(ctx, key, len);
+		if (lmd_res != 0) {
+			FMSG("hmac starts failed, res is 0x%x", -lmd_res);
+			return TEE_ERROR_GENERIC;
```

Comment:
As you suggest, I will add a helper function like this:
static TEE_Result mac_get_result_helper(int lmd_res)
{
	switch (lmd_res) {
	case MBEDTLS_ERR_MD_BAD_INPUT_DATA:
		return TEE_ERROR_BAD_PARAMETERS;
	case MBEDTLS_ERR_MD_ALLOC_FAILED:
		return TEE_ERROR_OUT_OF_MEMORY;
	case 0:
		return TEE_SUCCESS;
	default:
		return TEE_ERROR_BAD_STATE;
	}
}
Because the mbedtls error type is different for different algo, so we need different helper function for different algo.

Reply:
Where it makes sense, accurately reporting out of memory is one of the more important cases.